src/OrbitGl/OrbitApp.cpp

```diff
@@ -2291,35 +2291,54 @@ void OrbitApp::UpdateAfterSymbolLoading() {
   if (!HasCaptureData()) {
     return;
   }
-  const CaptureData& capture_data = GetCaptureData();
+  Future<std::shared_ptr<PostProcessedSamplingData>> post_processed_sampling_data_future =
+      thread_pool_->Schedule([this]() {
+        const CaptureData& capture_data = GetCaptureData();
 
-  PostProcessedSamplingData post_processed_sampling_data =
-      orbit_client_model::CreatePostProcessedSamplingData(capture_data.GetCallstackData(),
-                                                          capture_data, *module_manager_);
-  GetMutableCaptureData().set_post_processed_sampling_data(post_processed_sampling_data);
-  auto selection = std::make_unique<SelectionData>(*module_manager_, GetCaptureData(),
-                                                   GetCaptureData().post_processed_sampling_data(),
-                                                   &GetCaptureData().GetCallstackData());
-  main_window_->SetTopDownView(selection->GetTopDownView());
-  main_window_->SetBottomUpView(selection->GetBottomUpView());
-  main_window_->UpdateSamplingReport(&selection->GetCallstackData(),
-                                     &selection->GetPostProcessedSamplingData());
-  full_capture_selection_ = std::move(selection);
-
-  main_window_->ClearCallstackInspection();
-  inspection_selection_.reset();
-  time_range_thread_selection_.reset();
-
-  PostProcessedSamplingData selection_post_processed_sampling_data =
-      orbit_client_model::CreatePostProcessedSamplingData(capture_data.selection_callstack_data(),
-                                                          capture_data, *module_manager_);
-  GetMutableCaptureData().set_selection_post_processed_sampling_data(
-      std::move(selection_post_processed_sampling_data));
-
-  SetSelectionTopDownView(capture_data.selection_post_processed_sampling_data(), capture_data);
-  SetSelectionBottomUpView(capture_data.selection_post_processed_sampling_data(), capture_data);
-  main_window_->UpdateSelectionReport(&capture_data.selection_callstack_data(),
-                                      &capture_data.selection_post_processed_sampling_data());
+        return std::make_shared<PostProcessedSamplingData>(
+            orbit_client_model::CreatePostProcessedSamplingData(capture_data.GetCallstackData(),
+                                                                capture_data, *module_manager_));
+      });
+  Future<std::shared_ptr<std::unique_ptr<SelectionData>>> selection_future =
+      post_processed_sampling_data_future.Then(
+          thread_pool_.get(),
+          [this](const std::shared_ptr<PostProcessedSamplingData>& post_processed_sampling_data) {
+            GetMutableCaptureData().set_post_processed_sampling_data(
```

Comment:
But this still doesn't happen on the main thread, or does it? You seem to be using `thread_pool_.get()` as executor.

Reply:
Agreed. This should happen on the main thread. Is it possible to set this later and make it part of the return value?